### PR TITLE
Fix uppercase lowercase isses

### LINKED ISSUE #340: Environment variable names are converted to lower case
Environment variable names are converted to lower case internally. This is surprising undocumented behavior. We want to be able to override values from a configuration file with environment variables. It would be natural to use the same spelling (not counting the prefix), but this doesn't work if the configuration file keys are upper case. To make it work, the configuration file keys must be lower case.

config.toml
```bash
FOO="You can't override me!"
bar="You can override me just fine."
```
In the shell
```bash
APP_FOO="Why are you ignoring me?" APP_BAR="Yay! New bar." ./target/debug/app
```
I was using TOML for the config file and bash for the shell, if it matters.

### PR DESCRIPTION
**Pull request by YounessBird**
fix #340

### CODE REVIEW

**Comment by YounessBird** [comment]
Todo:
 json, json5, ini, ron

**Review by pictographer — APPROVED** [review]
Looks good.

One suggestion. Mention that keys are case insensitive in the docs.

**Comment by YounessBird** [comment]
Thanks. I will add your suggestion to the the chore list :)

File: README.md [hunk 1]
```diff
@@ -9,22 +9,24 @@
...
+-   Read from environment
+-   Loosely typed — Configuration values may be read in any supported type, as long as there exists a reasonable conversion
+-   Access nested fields using a formatted path — Uses a subset of JSONPath; currently supports the child ( `redis.port` ) and subscript operators ( `databases[0].name` )
+
+[json]: https://github.com/serde-rs/json
+[toml]: https://github.com/toml-lang/toml
+[yaml]: https://github.com/chyh1990/yaml-rust
+[ini]: https://github.com/zonyitoo/rust-ini
+[ron]: https://github.com/ron-rs/ron
+[json5]: https://github.com/callum-oakley/json5-rs
```
Comment (matthiasbeyer) [inline README.md:24]: Why was this changed?

File: README.md [hunk 2]
```diff
@@ -9,22 +9,24 @@
...
+
+Please note this library
+
+-   Can not be used to write changed configuration values back to the configuration file(s)!
+-   It's case insensitive and all the keys are converted to lowercase internally.
```
Comment (matthiasbeyer) [inline README.md:29]:
```suggestion
-   Is case insensitive and all the keys are converted to lowercase internally.
```

File: README.md [hunk 3]
```diff
@@ -35,12 +37,12 @@ config = "0.13.1"
...
- - `yaml` - Adds support for reading YAML files
- - `toml` - Adds support for reading TOML files
- - `ron` - Adds support for reading RON files
- - `json5` - Adds support for reading JSON5 files
+-   `ini` - Adds support for reading INI files
+-   `json` - Adds support for reading JSON files
+-   `yaml` - Adds support for reading YAML files
+-   `toml` - Adds support for reading TOML files
+-   `ron` - Adds support for reading RON files
+-   `json5` - Adds support for reading JSON5 files
```
Comment (matthiasbeyer) [inline README.md:45]: This is also an unrelated change.

File: src/config.rs [hunk 4]
```diff
@@ -149,6 +151,8 @@ impl Config {
     }
 
     pub fn get<'de, T: Deserialize<'de>>(&self, key: &str) -> Result<T> {
+        let k = key.to_lowercase();
+        let key = k.as_str();
```
Comment (matthiasbeyer) [inline src/config.rs:155]:
```suggestion
        let key = key.to_lowercase().as_str();
```

Does this work?

File: src/de.rs [hunk 5]
```diff
@@ -345,19 +345,8 @@ impl<'de> de::Deserializer<'de> for Config {
...
-            ValueKind::U64(i) => visitor.visit_u64(i),
-            ValueKind::U128(i) => visitor.visit_u128(i),
-            ValueKind::Boolean(b) => visitor.visit_bool(b),
-            ValueKind::Float(f) => visitor.visit_f64(f),
-            ValueKind::String(s) => visitor.visit_string(s),
-            ValueKind::Array(values) => visitor.visit_seq(SeqAccess::new(values)),
-            ValueKind::Table(map) => visitor.visit_map(MapAccess::new(map)),
-        }
+        // Delegate deserialization to Value
+        de::Deserializer::deserialize_any(self.cache, visitor)
```
Comment (matthiasbeyer) [inline src/de.rs:349]: How is this change related to the uppercase-lowercase issue?

File: src/lib.rs [hunk 6]
```diff
@@ -14,7 +14,7 @@
 //!  - Live watching and re-reading of configuration files
 //!  - Deep access into the merged configuration via a path syntax
 //!  - Deserialization via `serde` of the configuration or any subset defined via a path
-//!
+//! Please note that this crate is case insensitive and all the keys are converted to lowercase internally.
```
Comment (matthiasbeyer) [inline src/lib.rs:17]: I guess you can leave the blank line after the listing. AFAIK having a blank line after a listing is considered good style in markdown.

Care to change this? :laughing:

File: tests/Settings.json [hunk 7]
```diff
@@ -1,19 +1,21 @@
...
+        "latitude": 10.3970522,
+        "favorite": false,
+        "reviews": 3866,
+        "rating": 4.5,
+        "creator": {
+            "name": "John Smith",
+            "username": "jsmith",
+            "email": "jsmith@localhost"
+        }
+    },
```
Comment (matthiasbeyer) [inline tests/Settings.json:18]: AFAICS you only changed the indentation here. Care to revert this so we get a smaller diff?

File: tests/Settings.json5 [hunk 8]
```diff
@@ -1,20 +1,22 @@
...
+        latitude: 10.3970522,
+        favorite: false,
+        reviews: 3866,
+        rating: 4.5,
+        creator: {
+            name: "John Smith",
+            username: "jsmith",
+            email: "jsmith@localhost",
+        },
+    },
```
Comment (matthiasbeyer) [inline tests/Settings.json5:19]: Same as above.

File: tests/Settings.yaml [hunk 9]
```diff
@@ -2,13 +2,16 @@ debug: true
...
+    name: Torre di Pisa
+    longitude: 43.7224985
+    latitude: 10.3970522
+    favorite: false
+    reviews: 3866
+    rating: 4.5
+    creator:
+        name: John Smith
+        username: jsmith
+        email: jsmith@localhost
```
Comment (matthiasbeyer) [inline tests/Settings.yaml:14]: ...and here.

**Review by matthiasbeyer — CHANGES_REQUESTED** [review]
So first of all, thanks for working on this. This is important to get right.

I have some remarks, nothing too critical.

Still, I am not sure whether this is the right way to go. Converting all things to lowercase seems ... I don't know to be honest.

I am willing to merge this (modulo some things that have to be cleared), if we do not find another way... that said, when (if) we re-do the crate (#321) we should get it right from the get-go and support case sensitivity as well as insensitivity (let the user decide). But that's of course for another day, I guess.

File: README.md (hunk 1 above)
Comment (YounessBird) [reply README.md:24]: Unintentional change as I was trying to find the best place to add that the crate is case insensitive.

File: src/config.rs (hunk 4 above)
Comment (YounessBird) [reply src/config.rs:155]: No I got the  "temporary value dropped while borrowed (rust)" error.  If you know of a more elegant way I will be happy to amend it.

File: src/de.rs (hunk 5 above)
Comment (YounessBird) [reply src/de.rs:349]: Unrelated to the uppercase-lowercase issue. However, I felt it's an overkill to open an issue to point out this minor change, so while i was working on the same file I thought of introducing this change. If you feel like it's unnecessary or you prefer the old way I will be happy to remove it :) .

File: tests/Settings.json (hunk 7 above)
Comment (YounessBird) [reply tests/Settings.json:18]: there two more line that I have added to this file 
` "FOO": "FOO should be overridden",
    "bar": "I am bar"
 `
the extra space is introduced by running rustfmt

File: tests/Settings.json5 (hunk 8 above)
Comment (YounessBird) [reply tests/Settings.json5:19]: there two more line that I have added to this file 
` "FOO": "FOO should be overridden",
    "bar": "I am bar"
 `
the extra space is introduced by running rustfmt

File: tests/Settings.yaml (hunk 9 above)
Comment (YounessBird) [reply tests/Settings.yaml:14]: there two more line that I have added to this file 
` "FOO": "FOO should be overridden",
    "bar": "I am bar"
 `
the extra space is introduced by running rustfmt

File: README.md (hunk 1 above)
Comment (matthiasbeyer) [reply README.md:24]: No I think you changed some whitespace here. For example in c767d28 and in 0a2e8eb I can see whitespace change.

Sorry for being so persistent on this, I just want to be sure that the diff only touches what needs to be touched, because a larger diff makes harder to review. Right now it is not that much, but I hope you see the point: If a PR runs for weeks and the diff gets more and more overhead, this results in a discouraging amount of review work :laughing: 

If you don't want to continue because of my persistence, I can fully understand this and, if you're okay, I can take over this PR!

File: src/de.rs (hunk 5 above)
Comment (matthiasbeyer) [reply src/de.rs:349]: >  you feel like it's unnecessary or you prefer the old way I will be happy to remove it :) .

If you don't mind, feel free to open a seperate PR for this and remove it from this PR.

If not, I can do this myself if you like! :laughing:

File: src/lib.rs (hunk 6 above)
Comment (matthiasbeyer) [reply src/lib.rs:17]: Oh okay, interesting. Can you link me to the CI failure for this?

File: tests/Settings.json (hunk 7 above)
Comment (matthiasbeyer) [reply tests/Settings.json:18]: rustfmt changes a JSON file?

File: README.md (hunk 1 above)
Comment (YounessBird) [reply README.md:24]: > No I think you changed some whitespace here. For example in https://github.com/mehcode/config-rs/commit/c767d285d694f43539e4e98b6580024644cab22a and in https://github.com/mehcode/config-rs/commit/0a2e8ebcfe3951ea33ab99a3782f9193dd87d9d3 I can see whitespace change.

Yes I see what you mean. I think that has to do with the extension I am using when I hit save, my bad.

> Sorry for being so persistent on this, I just want to be sure that the diff only touches what needs to be touched, because a larger diff makes harder to review. Right now it is not that much, but I hope you see the point...

You are absolutely right, and at the time when I pushed this I just didn't notice these extra unnecessary changes in my commit. Thank you for pointing them out, I am gonna try and amend them later.

File: src/de.rs (hunk 5 above)
Comment (YounessBird) [reply src/de.rs:349]: Yes I will make sure to remove this change from this pr :).

File: tests/Settings.json (hunk 7 above)
Comment (YounessBird) [reply tests/Settings.json:18]: Sorry I think that has to do with the extension I am using when I hit save, my bad.

File: src/lib.rs (hunk 6 above)
Comment (YounessBird) [reply src/lib.rs:17]: Yes, I think is this [one](https://github.com/mehcode/config-rs/actions/runs/2615343121)

Comment (matthiasbeyer) [reply src/lib.rs:17]: Ah. Yes, I see what gave you that impression: If you look closely, you'll see that on line 11 of the log you linked, the line is removed, but on line 12 it is added again.

This is usually an indicator of trailing whitespace that was removed! So there was a trailing whitespace that got removed, which shows as the line removed (line 11) and it looks like the very same line was added (line 12), but it is one without a trailing whitespace.
Unfortunately it is not clearly visible in the log.

I don't mind this, so I will mark this as resolved. :laughing:

**Review by matthiasbeyer — APPROVED** [review]
This was laying around way too long. I'm gonna merge this now, sorry for me taking so long!

**Comment by utkarshgupta137** [comment]
There is still a bug (as of 0.13.3).
Consider this:
```
let conf: HashMap<String, String> = Config::builder()
    .add_source(File::with_name(name))
    .add_source(Environment::with_prefix(prefix))
    .build()?
    .try_deserialize()?;
```
File:
```
KEY = "value1"
```
Env:
```
PREFIX_KEY=value2
```

Then:
Current behavior: conf has 2 keys: `KEY` & `key` with values `value1` & `value2`
Expected behavior: conf has 1 key: `KEY` with value `value2`
